Create background events for URL validation

johnwatkins0 opened the pull request:
## Summary

Fixes #1756. Some related discussion also took place in #5228

### Main changes

This PR does the following:

- Creates an ~hourly~ _daily_ cron task that validates a set of URLs. The URLs are the default set returned from the `ScannableURLProvider` class. This set will always include most recent posts for all AMP-enabled post types plus a standard set of others such as the homepage and a search result.
- Triggers a single background event on `save_post` to validate the saved post's URL when the user has dev tools off. When the user has dev tools on, the existing behavior (URL is validated synchronously on save) is unchanged.

### Additional changes

#### BackgroundTaskDeactivator
I also took the suggestion from this to-do comment in the `CronBasedBackgroundTask` class:

> @todo Needs refactoring if used for more than one cron-based task, to avoid iterating over sites multiple times.

I created the new class, `BackgroundTaskDeactivator` which is injected into the background task classes and used to queue up registered background tasks to be cleared when the plugin is deactivated. 


#### CronBasedBackgroundTask refactor

`CronBasedBackgroundTask` was previously being extended for a couple of different recurring cron tasks. I have created `RecurringBackgroundTask` along with `SingleScheduledBackgroundTask` to account for the two different scheduling approaches we need for background. These new classes are both abstract, inheriting from `CronBasedBackgroundTask`, and `CronBasedBackroundTask` is no longer directly extended by any non-abstract class.

#### Remove locking

I removed the locking functions from `URLValidationProvider`. I introduced these in another recent PR, but we subsequently determined they're not needed for now at least.

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

johnwatkins0 committed:
Create single scheduled background task with save_post single event

johnwatkins0 committed:
Additional work on SavePostValidationEvent and its parent class

johnwatkins0 committed:
Include cli utils in autoload-dev

johnwatkins0 committed:
Remove cli utils from autoload-dev

johnwatkins0 committed:
Merge branch 'feature/refactor-url-validation' into feature/1756-validation-cron

johnwatkins0 committed:
Increase unit test coverage for new classes

johnwatkins0 committed:
Undo changes to bootstrap file

johnwatkins0 committed:
Add newline to end of file

johnwatkins0 committed:
Fix tests for MonitorCssTransientCaching

johnwatkins0 committed:
Fix tests for ValidatedUrlStylesheetDataGarbageCollection

johnwatkins0 committed:
Clean up SingleScheduledBackgroundTask

johnwatkins0 committed:
SavePostValidationEvent cleanup

johnwatkins0 committed:
Clean up and simplify tests

johnwatkins0 committed:
Remove unused class property

johnwatkins0 committed:
PHPCS and other cleanup

johnwatkins0 committed:
Increase test coverage

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=5515) <br/>All committers have signed the CLA.

johnwatkins0 committed:
Merge branch 'feature/refactor-url-validation' into feature/1756-validation-cron

johnwatkins0 committed:
PHPStan fixes

johnwatkins0 committed:
Pass force validation flags in cron callbacks

johnwatkins0 committed:
Fix a since version

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

johnwatkins0 committed:
Remove use of obsolete class constant

johnwatkins0 committed:
Integrate updated ValidationRequestMocking (converted to trait)

johnwatkins0 committed:
Comment updates and minor refactors

johnwatkins0 commented:
Requesting review on this.

johnwatkins0 commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,116 @@
+
+		$validation_provider = new URLValidationProvider();
+
+		$validation_provider->with_lock(
```
@westonruter Here's the next use case for `with_lock` outside the CLI script. What do you think? Is it fine if, in some rare scenarios, this cron task validates a bunch of URLs at the same time someone is running the CLI script? If so, I'll remove the locking. 

My thinking was that having multiple validation processes running simultaneously could overload some servers. But on the other hand, by default we're only validating a few URLs at a time, and sleeping between each request. Anyone with the knowledge to validate a lot of URLs at a time can do so at their own risk.

westonruter replied:
I think that it is going to be very unlikely that people will be often using the CLI script to do validation. Most likely the vast majority of sites that have limited resources are shared servers on which hosts don't even offer WP-CLI access in the first place. So I think that the locking isn't necessary right now, but it's something we can keep in mind in the future if need be.

johnwatkins0 replied:
Sounds good. I'll update.

johnwatkins0 replied:
Removed locking in 2d2a0a62b

pierlon commented on src/BackgroundTask/BackgroundTaskDeactivator.php:
```diff
@@ -0,0 +1,171 @@
+			foreach ( get_sites(
+				[
+					'fields' => 'ids',
+					'number' => 0,
```
Setting `number` to 0 means there wouldn't be any results, no?

johnwatkins0 replied:
https://core.trac.wordpress.org/browser/tags/5.5.1/src/wp-includes/class-wp-site-query.php#L447

If I'm understanding this correctly, setting this to 0 prevents pagination in the query and retrieves all sites. Do you agree with this reading?

This code was originally written by @schlessera before I refactored, so he probably has a clear answer.

schlessera replied:
Yes, this fetches all sites without pagination. The name of the argument is misleading, unfortunately. Might be good to add a clarifying comment for our future selves.

Note: the unpaginated call is practically limited still as we are checking for a large site first.

johnwatkins0 replied:
Added a comment in b08f1649b

pierlon commented on src/BackgroundTask/BackgroundTaskDeactivator.php:
```diff
@@ -0,0 +1,171 @@
+			return $actions;
+		}
+
+		wp_enqueue_style( 'amp-icons' );
+
+		$warning_icon = $this->get_warning_icon();
+		if ( false === strpos( $actions['deactivate'], $warning_icon ) ) {
+			$actions['deactivate'] = preg_replace( '#(?=</a>)#i', ' ' . $warning_icon, $actions['deactivate'] );
+		}
```
@westonruter I noticed you were the original author of this snippet of code.

How about instead of only appending the warning icon to the anchor text, we use a confirmation dialog to verify they want to deactivate? For example:

```suggestion
		if ( preg_match( '/<a\s+.*?id\s*=\s*"([^"]+)"/', $actions['deactivate'], $matches ) ) {
			$anchor_id = $matches[1];
			$confirm_text = __( 'Deactivation will leave orphaned scheduled events behind. Do you want to continue?', 'amp' );

			$script = "
				document.getElementById( '{$anchor_id}' ).addEventListener( 'click', function ( event ) {
					if ( ! confirm( '{$confirm_text}' ) ) {
						event.preventDefault();
					}
				} )
			";

			wp_add_inline_script( 'wp-dom-ready', $script );
		}
```

And here's it in action:

![](https://user-images.githubusercontent.com/16200219/100193984-ceb41500-2eec-11eb-8699-88423e4b8e22.gif)

westonruter replied:
Actually, the warning was originally added by @schlessera in eaf8259b87cc76a0782599ebea5583d8e5d143d2, although it was further iterated on.

Adding a `confirm()` would work. The only downside would be if you select multiple plugins and do a bulk deactivation: you'd easily miss that the plugin has special considerations for deactivation.

johnwatkins0 replied:
I have to say I'm pretty agnostic on this. We could do both (the icon and the confirm), make no change, or make a new issue for further discussion. Which way are you leaning @westonruter @pierlon ?

pierlon replied:
> The only downside would be if you select multiple plugins and do a bulk deactivation: you'd easily miss that the plugin has special considerations for deactivation.

@westonruter The `confirm()` could also be applied to that also, no?

I agree with @johnwatkins0 with regards to moving this discussion to its own issue.

pierlon commented on src/BackgroundTask/BackgroundTaskDeactivator.php:
```diff
@@ -0,0 +1,171 @@
+	 * @param string   $plugin_file Path to the plugin file relative to the plugins directory.
+	 * @return string[]
+	 */
+	public function add_warning_to_plugin_meta( $plugin_meta, $plugin_file ) {
```
@westonruter Same suggestion as above. If you do plan to do something similar to above, this method may not be needed. Instead, you could probably hook into `plugin_action_links_{$this->plugin_file}`.

pierlon commented on src/BackgroundTask/BackgroundTaskDeactivator.php:
```diff
@@ -0,0 +1,171 @@
+	 * Class constructor.
+	 */
+	public function __construct() {
+		$this->plugin_file = plugin_basename( AMP__DIR__ . '/amp.php' );
```
```suggestion
		$this->plugin_file = plugin_basename( AMP__FILE__ );
```

pierlon commented on src/BackgroundTask/SingleScheduledBackgroundTask.php:
```diff
@@ -0,0 +1,112 @@
+ * @since 2.1
+ * @internal
+ */
+abstract class SingleScheduledBackgroundTask implements Service, Registerable, Conditional {
```
Ideally, this should be extending `CronBasedBackgroundTask`, but the main blocker is the  `process` abstract method not being able to be overridden (which has only been possible [since PHP  7.2](https://wiki.php.net/rfc/allow-abstract-function-override)).

One possible solution could be to update the `process` method in `CronBasedBackgroundTask` to match the same arguments in this class, but I think what's here is just fine for now.

johnwatkins0 replied:
I made a brief attempt at this, but it became complicated when it came to `MonitorCssTransientCaching`; I was going to have to make modifications to that class to accommodate the signature changes to `CronBasedBackgroundTask`. I think we should leave this alone for now. Using class extension might be good for DRY, but cron tasks are already difficult to debug as it is, so it might be wise to keep these classes straightforward, even if some code is repeated. I'm happy to revisit, though, if others disagree.

schlessera replied:
Note that the initial abstraction was more flexible, you can still see it here: https://github.com/ampproject/amp-wp/pull/4177/commits/4d2bdba6e40c7d6738bb9f3afb71c4b66cef1d1d

It was later simplified, as that kind of abstraction was not really needed, with the intention of adapting it once there is an actual need.

pierlon commented on tests/php/src/ValidatedUrlStylesheetDataGarbageCollectionTest.php:
```diff
@@ -5,6 +5,7 @@
  * @package AmpProject\AmpWP
  */
 
```
The namespace is missing along with a few classes to be imported. Suggested patch:

```diff
diff --git a/tests/php/src/ValidatedUrlStylesheetDataGarbageCollectionTest.php b/tests/php/src/ValidatedUrlStylesheetDataGarbageCollectionTest.php
--- a/tests/php/src/ValidatedUrlStylesheetDataGarbageCollectionTest.php	(revision cd4dd777102b5f30b32728e0c40cfaecc93b4097)
+++ b/tests/php/src/ValidatedUrlStylesheetDataGarbageCollectionTest.php	(date 1606295527974)
@@ -5,8 +5,12 @@
  * @package AmpProject\AmpWP
  */
 
+namespace AmpProject\AmpWP\Tests;
+
+use AMP_Validated_URL_Post_Type;
 use AmpProject\AmpWP\BackgroundTask\BackgroundTaskDeactivator;
 use AmpProject\AmpWP\BackgroundTask\ValidatedUrlStylesheetDataGarbageCollection;
+use WP_UnitTestCase;
 
 /** @coversDefaultClass \AmpProject\AmpWP\BackgroundTask\ValidatedUrlStylesheetDataGarbageCollection */
 class ValidatedUrlStylesheetDataGarbageCollectionTest extends WP_UnitTestCase {
```

johnwatkins0 replied:
Done in 28581d774.

pierlon commented on tests/php/src/Validation/SavePostValidationEventTest.php:
```diff
@@ -0,0 +1,65 @@
+use WP_UnitTestCase;
+
+/**
+ * @coversDefaultClass \AmpProject\AmpWP\Validation\SavePotValidationEvent
```
```suggestion
 * @coversDefaultClass \AmpProject\AmpWP\Validation\SavePostValidationEvent
```

pierlon commented on tests/php/src/Validation/URLValidationCronTest.php:
```diff
@@ -0,0 +1,68 @@
+use AmpProject\AmpWP\Validation\URLValidationCron;
+use WP_UnitTestCase;
+
+/** @coversDefaultClass AmpProject\AmpWP\Validation\URLValidationCron */
```
```suggestion
/** @coversDefaultClass \AmpProject\AmpWP\Validation\URLValidationCron */
```

pierlon commented on tests/php/test-class-monitor-css-transient-caching.php:
```diff
@@ -5,6 +5,7 @@
  * @package AmpProject\AmpWP
  */
 
```
This test should be moved to the `AmpProject\AmpWP\Tests` namespace. Suggested patch:

```diff
diff --git a/tests/php/test-class-monitor-css-transient-caching.php b/tests/php/src/BackgroundTask/MonitorCssTransientCachingTest.php
rename from tests/php/test-class-monitor-css-transient-caching.php
rename to tests/php/src/BackgroundTask/MonitorCssTransientCachingTest.php
--- a/tests/php/test-class-monitor-css-transient-caching.php	(revision cd4dd777102b5f30b32728e0c40cfaecc93b4097)
+++ b/tests/php/src/BackgroundTask/MonitorCssTransientCachingTest.php	(date 1606295939069)
@@ -1,20 +1,18 @@
 <?php
 /**
- * Class Test_Monitor_CSS_Transient_Caching.
- *
- * @package AmpProject\AmpWP
+ * Tests for the MonitorCssTransientCaching class.
  */
 
+namespace AmpProject\AmpWP\Tests\BackgroundTask;
+
+use AMP_Options_Manager;
 use AmpProject\AmpWP\BackgroundTask\BackgroundTaskDeactivator;
 use AmpProject\AmpWP\BackgroundTask\MonitorCssTransientCaching;
 use AmpProject\AmpWP\Option;
+use WP_UnitTestCase;
 
-/**
- * Test MonitorCssTransientCaching.
- *
- * @coversDefaultClass \AmpProject\AmpWP\BackgroundTask\MonitorCssTransientCaching
- */
-class Test_Monitor_CSS_Transient_Caching extends WP_UnitTestCase {
+/** @coversDefaultClass \AmpProject\AmpWP\BackgroundTask\MonitorCssTransientCaching */
+class MonitorCssTransientCachingTest extends WP_UnitTestCase {
 
 	/**
 	 * Whether external object cache is being used.
```

johnwatkins0 replied:
Done in 3d07f78d6

pierlon replied:
Only thing left to do here is to rename the file to `tests/php/src/BackgroundTask/MonitorCssTransientCachingTest.php`.

pierlon replied:
Also, a small patch to be applied:

```diff
diff --git a/tests/php/test-class-monitor-css-transient-caching.php b/tests/php/test-class-monitor-css-transient-caching.php
--- a/tests/php/test-class-monitor-css-transient-caching.php	(revision 7cdde967dd63af7b7b1d8f945efe1a1060d55662)
+++ b/tests/php/test-class-monitor-css-transient-caching.php	(date 1608718679528)
@@ -45,7 +45,7 @@
 	 * Test whether an event is actually scheduled when the monitor is registered.
 	 *
 	 * @uses \AmpProject\AmpWP\BackgroundTask\CronBasedBackgroundTask::schedule_event
-	 * @uses \AmpProject\AmpWP\BackgroundTask\CronBasedBackgroundTask::deactivate
+	 * @uses \AmpProject\AmpWP\BackgroundTask\BackgroundTaskDeactivator::deactivate
 	 */
 	public function test_event_gets_scheduled_and_unscheduled() {
 		wp_set_current_user( self::factory()->user->create( [ 'role' => 'administrator' ] ) );
```

johnwatkins0 replied:
Done in 1a4043156

johnwatkins0 committed:
Remove URL validation locking

johnwatkins0 committed:
Simplify plugin basename retrieval

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Remove unused variable

johnwatkins0 committed:
Merge branch 'feature/1756-validation-cron' of https://github.com/ampproject/amp-wp into feature/1756-validation-cron

johnwatkins0 committed:
Add namespace to ValidatedUrlStylesheetDataGarbageCollectionTest.php

johnwatkins0 committed:
Typo fix

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Add leading slash to coversDefaultClass comment

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Add namespacing to MonitorCssTransientCachingTest.php

johnwatkins0 committed:
Merge branch 'feature/1756-validation-cron' of https://github.com/ampproject/amp-wp into feature/1756-validation-cron

johnwatkins0 committed:
Increase test coverage

westonruter commented on src/BackgroundTask/CronBasedBackgroundTask.php:
```diff
@@ -20,20 +18,20 @@
 	 */
-	private $plugin_file;
+	public function __construct( BackgroundTaskDeactivator $background_task_deactivator ) {
+		$background_task_deactivator->add_event( $this->get_event_name() );
```
Would this be better suited to be called in the `register` method?

johnwatkins0 replied:
Good call. Done in 5334bdf62

westonruter commented on src/BackgroundTask/SingleScheduledBackgroundTask.php:
```diff
@@ -0,0 +1,112 @@
+	 * @param BackgroundTaskDeactivator $background_task_deactivator Service that deactivates background events.
+	 */
+	public function __construct( BackgroundTaskDeactivator $background_task_deactivator ) {
+		$background_task_deactivator->add_event( $this->get_event_name() );
```
Wouldn't this make more sense to be called in the `register` method?

westonruter replied:
Or, come to think of it, there could be a default `register` implementation that does `$this->background_task_deactivator->add_event( $this->get_event_name() )`. The constructor could be just responsible for storing `$background_task_deactivator` as a member variable.

westonruter commented on src/BackgroundTask/SingleScheduledBackgroundTask.php:
```diff
@@ -0,0 +1,112 @@
+	/**
+	 * Register the service with the system.
+	 */
+	public function register() {
```
Per above, this could then do:

```suggestion
	public function register() {
		parent::register();
```

johnwatkins0 replied:
I think I'll need to make `SingleScheduledBackgroundTask` and `CronBasedBackgroundTask` inherit from a shared parent class before I can do this.

johnwatkins0 replied:
Done in 94c99599d

westonruter commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -0,0 +1,139 @@
+	public function process( ...$args ) {
+		$post_id = reset( $args );
+
+		if ( empty( get_post( $post_id ) ) ) {
```
Since `$post_id` being `0` can result in `get_post()` returning the global `$post`.

```suggestion
		if ( empty( $post_id ) || empty( get_post( $post_id ) ) ) {
```

westonruter commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -0,0 +1,139 @@
+
+		$id = reset( $args );
+
+		if ( wp_is_post_revision( $id ) ) {
```
There are other checks that should be done here I think:

https://github.com/ampproject/amp-wp/blob/8a95f4b8d9b11f24ae2e586d556e650eb64c52ee/includes/validation/class-amp-validation-manager.php#L241-L254

johnwatkins0 replied:
Done in 8fefced7e

westonruter commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -0,0 +1,139 @@
+	 */
+	private function get_url_validation_provider() {
+		if ( is_null( $this->url_validation_provider ) ) {
+			$this->url_validation_provider = new URLValidationProvider();
```
Shouldn't this be using dependency injection? Based on @schlessera's guidance, see https://github.com/ampproject/amp-wp/pull/5558/commits/d1e87320d1beb216e8059ca496980f3b187d4bc7. This is part of my as-of-yet unmerged #5558 PR.

johnwatkins0 replied:
Interesting. My (admittedly incomplete) understanding was that we were using dependency injection for shared instances, but not for objects that are not necessarily shared. But following your commit and looking at `SimpleInjector::make`, I see there is more to it than that. I will update.

johnwatkins0 replied:
Done in 16e3ee53b37ca2d6d1abedadd9d12705b8bc3a68 and e2fe6cda9633ad96558b5ed8b30340d8ffb488ed

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,112 @@
+	 */
+	public function process() {
+		$number_per_type         = $this->get_url_validation_number_per_type();
+		$validation_url_provider = new ScannableURLProvider( $number_per_type, [], true );
```
Ditto above about using the injector.

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,112 @@
+		$urls                    = $validation_url_provider->get_urls();
+		$sleep_time              = $this->get_sleep_time();
+
+		$validation_provider = new URLValidationProvider();
```
Ditto using injector.

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,112 @@
+		 *
+		 * @param int The number of URLs. Default 1.
+		 */
+		return (int) apply_filters( 'amp_url_validation_number_per_type', self::DEFAULT_LIMIT_PER_TYPE );
```
What if someone filters this to be `-1` or `0`?

johnwatkins0 replied:
Updated in 69eddd7e6 to ensure the number is -1 or above, but do you think we should forbid `-1` to prevent validating all posts? `0` should be fine because it would just make the process do nothing.

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,112 @@
+		 *
+		 * @param int The number of seconds. Default 1.
+		 */
+		return (int) apply_filters( 'amp_url_validation_sleep_time', self::DEFAULT_SLEEP_TIME );
```
What if someone filters this to be `0` or a negative number?

johnwatkins0 replied:
Updated to use `absint` in 69eddd7e6. I think `0` should be allowed for cases where the server can easily handle it.

johnwatkins0 committed:
Harden post_id check

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Add comment to potentially misleading number arg in get_sites

johnwatkins0 committed:
Move background task deactivate add_event call to register methods

johnwatkins0 committed:
Add additional should_schedule_event checks

johnwatkins0 committed:
Use dependency injector to create instances

johnwatkins0 committed:
Add missing doc comment for parameter

johnwatkins0 committed:
Fix failing get_validated_urls test

johnwatkins0 committed:
Add injector delegation

johnwatkins0 committed:
Improve number validation

johnwatkins0 committed:
PHPStan param fix

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

johnwatkins0 committed:
Make SavePostValidationEvent inherit from CronBasedBackgroundTask

johnwatkins0 committed:
Create RecurringBackgroundTask

johnwatkins0 committed:
Use get_timestamp instead of get_interval in SingleScheduledBackgroundEvent

johnwatkins0 committed:
Add tests for schedule_event

johnwatkins0 committed:
Increase coverage for single scheduled event

johnwatkins0 committed:
Fix failing schedule_event test

johnwatkins0 commented:
@pierlon @schlessera @westonruter Re-requesting review on this. I've updated the PR description. This includes the refactor of the `CronBasedBackgroundEvents` class. I gave it two child classes, one for recurring events, and one for single events. It wasn't too complicated. Let me know if you see room for improvement.

johnwatkins0 committed:
Re-enable tests

johnwatkins0 committed:
Update test to use a different filter; other was since 5.2

johnwatkins0 committed:
Make anonymous functions static

johnwatkins0 committed:
Add more SavePostValidationEvent tests

johnwatkins0 committed:
Increase test coverage

johnwatkins0 committed:
Fix mistaken assertTrue instead of assertFalse

johnwatkins0 committed:
Add additional tests for get_url_validation_number_per_type

schlessera commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -0,0 +1,163 @@
+	 * @param Injector                  $injector Injector instance.
+	 * @param UserAccess                $dev_tools_user_access Dev tools user access class instance.
+	 */
+	public function __construct( BackgroundTaskDeactivator $background_task_deactivator, Injector $injector, UserAccess $dev_tools_user_access ) {
```
Is there a specific reason why you're injecting the injector here, and not just the object instance you actually need, i.e. the `URLValidationProvider` ?

schlessera commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,142 @@
+	 * @param BackgroundTaskDeactivator $background_task_deactivator Service that deactivates background events.
+	 * @param Injector                  $injector Injector instance.
+	 */
+	public function __construct( BackgroundTaskDeactivator $background_task_deactivator, Injector $injector ) {
```
Again, instead of injecting the injector, why not immediately injecting `ScannableURLProvider` and `URLValidationProvider` ?

johnwatkins0 replied:
Both of these classes are not necessarily shared services. Multiple instances with different options may be created in different places within a single execution (although I don't think that we're doing that anywhere right now outside testing). What's the recommended approach for creating non-shared instances (potentially with different options passed into their constructors) with our dependency injection?

schlessera replied:
Nothing is shared by the injector by default, every injection gets its own separate instance.

If you want to have an instance shared across all injections, it needs to be added to the list of shared instances:

https://github.com/ampproject/amp-wp/blob/2.0.8/src/AmpWpPlugin.php#L145-L165

schlessera replied:
So, in this case, just inject the services where needed, and it will just work, no need for injecting the injector.

johnwatkins0 replied:
All right, so if I add `ScannableURLProvider` to the constructor here, how would I pass arguments into `ScannableURLProvider`'s constructor when it's instantiated?

westonruter replied:
Like this? `$this->injector->make( ScannableURLProvider::class, [ /* arguments */ ] )`?

johnwatkins0 replied:
@westonruter Alain is suggesting that I not use the injector directly but instead declare `ScannableURLProvider` as a dependency in the constructor. So what I'm unsure of is, if I have,

```
public function __construct( BackgroundTaskDeactivator $background_task_deactivator, ScannableURLProvider $scannable_url_provider ) {
```

how do I pass arguments into `$scannable_url_provider`'s constructor?

I'll dig into the dependency injection code to see if there's an answer built in.

schlessera replied:
There are a few ways to solve this.
Let's summarize the constraints first (to make sure I understand correctly):
- `ScannableURLProvider` needs 3 arguments that differ between instantiations.
- These arguments need to be decided "late" as they depend on a filter, and we don't want to call that filter too early.

a. Inject a `ScannableURLProviderFactory` instead of a `ScannableURLProvider`. So, instead of injecting the entire Injector (which gives too much knowledge and power to the object that receives it, and couples it to knowledge about technical implementation details that are irrelevant to its business domain), we inject the possibility to decide on what and when to instantiate regarding the `ScannableURLProvider` only.

b. Turn the constructor arguments into an injectable object, like `URLScanningContext`. This makes it controllable again by the injector, and the injector can then include logic that decides what to inject. In this case, though, it is enough to just not have any extra logic in the injector at all. All we need is to lazily provide the needed settings only at the point where they are actually requested. The `URLScanningContext` could at that point run the filter transparently.

c. Move the constructor arguments from the constructor into the public method that triggers the action. So, instead of having 3 constructor arguments and 1 method arguments, you're left with 4 method arguments.

All 3 of the above should work. I think my preference is with b., as it cleanly separates the logic into meaningful bits and deals with the timing issue (which happens to be a WordPress technical detail) in a transparent way.

Feel free to ask questions about the above, or let me know if you want to chat in a call to discuss in more detail.

schlessera replied:
Hmm, looking back at the code, I think b. would not solve the third constraint I failed to recognize:
- The `URLValidationCron` wants to control the 3 constructor arguments when it runs and control the knowledge about them.

I think this requirement is not necessarily one I'd enforce, as it seems that has nothing to do with a cron job, but is more about resource allocation of the background scanning. So, I'd be fine with moving `get_url_validation_number_per_type` into the `URLScanningContext` instead. However, if you want to avoid that, you can go with a. above instead.

johnwatkins0 replied:
I went with option b and added `URLScanningContext` in 6747c04f03f0d480d71567ab6ddf165e9375eecf. 

I think in the future it may be necessary to add some setters to `URLScanningContext` to allow changing the configuration later in the process. For now the defaults work for our use cases, except in the CLI script, which is still outside the DI architecture so creates its own instances on the fly.

schlessera commented on src/BackgroundTask/BackgroundTaskDeactivator.php:
```diff
@@ -0,0 +1,171 @@
+ * @since 2.1
+ * @internal
+ */
+final class BackgroundTaskDeactivator implements Service, Conditional, Registerable, Deactivateable {
```
I think the `BackgroundTaskDeactivator` can be added to the list of _shared_ services, as we'd only need a shared single instance across all cron tasks for it to work.

johnwatkins0 replied:
Are `AmpWpPlugin::SERVICES` not shared instances? I thought they were, but it's possible I have been misunderstanding that. I think there are probably other Service classes I've created on other PRs that only need to be singletons.

schlessera replied:
No, they are not shared by default. See comment above.

schlessera replied:
Oh, wait, I'm referring to regular dependencies, not services. Let me check the injector again before I continue talking nonsense...

schlessera replied:
No, it seems correct, services are not shared by default.

johnwatkins0 replied:
Done in f38297f59

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

johnwatkins0 committed:
Move BackgroundTaskDeactivator to shared instances

johnwatkins0 committed:
Add URLScanningContext class

johnwatkins0 committed:
Fix failing URLValidationCronTest tests

johnwatkins0 commented:
@schlessera @pierlon @westonruter I have addressed all feedback aside from the open question about the network admin (which can be a separate issue if we do anything with that). This is ready for further review.

pierlon commented on src/Validation/URLScanningContext.php:
```diff
@@ -0,0 +1,105 @@
+	public function get_limit_per_type() {
+		/**
+		 * Filters the number of URLs per content type to check during each run of the cron task.
+		 *
```
We should be noting at what version a new filter is introduced, right @westonruter?

```suggestion
		 *
		 * @since 2.1.0
```

pierlon commented on src/Validation/URLScanningContext.php:
```diff
@@ -0,0 +1,105 @@
+		 *
+		 * @param int $url_validation_number_per_type The number of URLs. Default 1.
+		 */
+		$url_validation_limit_per_type = apply_filters( 'amp_url_validation_limit_per_type', $this->limit_per_type );
+
+		// Valid values are any integer -1 and above.
+		if ( -1 > $url_validation_limit_per_type ) {
+			return 1;
+		}
+
+		return $url_validation_limit_per_type;
```
```suggestion
		$url_validation_limit_per_type = (int) apply_filters( 'amp_url_validation_limit_per_type', $this->limit_per_type );

		return max( $url_validation_limit_per_type, 1 );
```

johnwatkins0 replied:
-1 is allowed, though, to scan all URLs. 0 is also technically allowed (could be used to disable the feature entirely).

westonruter replied:
I've refined this a bit more in f64e4ab93.

pierlon commented on src/Validation/URLScanningContext.php:
```diff
@@ -0,0 +1,105 @@
+	 */
+	public function get_limit_per_type() {
+		/**
+		 * Filters the number of URLs per content type to check during each run of the cron task.
```
Maybe this could be a bit more specific since it's going to be used as the description in the docs.

```suggestion
		 * Filters the number of URLs per content type to check during each run of the validation cron task.
```

pierlon commented on src/BackgroundTask/MonitorCssTransientCaching.php:
```diff
@@ -92,23 +92,25 @@ protected function get_event_name() {
+	 * @param mixed[] ...$args {
+	 *     Arguments passed to the cron callback.
+	 *
+	 *     @type DateTimeInterface @date Optional. Date to use for timestamping the processing (for testing).
```
```suggestion
	 *     @type DateTimeInterface $date Optional. Date to use for timestamping the processing (for testing).
```

pierlon commented on src/BackgroundTask/RecurringBackgroundTask.php:
```diff
@@ -0,0 +1,55 @@
+			return;
+		}
+
+		wp_schedule_event( time(), $this->get_interval(), $event_name );
```
Shouldn't `$args` also be passed to the scheduled event (although in this case the `admin_init` action will always return `[]`)?

```suggestion
		wp_schedule_event( time(), $this->get_interval(), $event_name, $args );
```

johnwatkins0 replied:
It can't hurt, and allows me to get rid of the phpcs:ignore above. Done in 1419ef4e3

pierlon commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,118 @@
+	 *
+	 * @param BackgroundTaskDeactivator $background_task_deactivator Service that deactivates background events.
+	 * @param ScannableURLProvider      $scannable_url_provider ScannableURLProvider instance.
+	 * @param URLValidationProvider     $url_validation_provider URLValidationProvider isntance.
```
```suggestion
	 * @param URLValidationProvider     $url_validation_provider URLValidationProvider instance.
```

pierlon commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,118 @@
+
+		/**
+		 * Filters the length of time to sleep between validating URLs.
+		 *
```
Similarly above, we should include the version when a filter is added for future reference.

```suggestion
		 *
		 * @since 2.1
```

pierlon commented on tests/php/src/Validation/SavePostValidationEventTest.php:
```diff
@@ -0,0 +1,162 @@
+use AmpProject\AmpWP\BackgroundTask\BackgroundTaskDeactivator;
+use AmpProject\AmpWP\BackgroundTask\SingleScheduledBackgroundTask;
+use AmpProject\AmpWP\DevTools\UserAccess;
+use AmpProject\AmpWP\Infrastructure\Injector\SimpleInjector;
```
This class seems to no longer be in use.

```suggestion
```

pierlon commented on tests/php/src/Validation/SavePostValidationEventTest.php:
```diff
@@ -0,0 +1,162 @@
+
+	/**
+	 * @covers ::process()
+	 * @covers ::get_url_validation_provider
```
Method doesn't exist on tested class.

```suggestion
```

pierlon commented on tests/php/src/Validation/URLScanningContextTest.php:
```diff
@@ -0,0 +1,81 @@
+		$this->assertInstanceof( URLScanningContext::class, $this->test_instance );
+	}
+
+	/** @covers ::@covers ::get_url_validation_limit_per_type() */
+	public function test_get_amp_url_validation_limit_per_type() {
```
```suggestion
	/** @covers ::get_limit_per_type() */
	public function test_get_limit_per_type() {
```

pierlon commented on tests/php/src/Validation/URLValidationCronTest.php:
```diff
@@ -0,0 +1,120 @@
+	 * Test validate_urls.
+	 *
+	 * @covers ::process()
+	 * @covers ::validate_urls()
```
Not a method on tested class.

```suggestion
```

johnwatkins0 committed:
Fix typo in type comment

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Pass args to wp_schedule_event

johnwatkins0 committed:
Merge branch 'feature/1756-validation-cron' of https://github.com/ampproject/amp-wp into feature/1756-validation-cron

johnwatkins0 committed:
Minor updates to comments and unnecessary use statements

Co-authored-by: Pierre Gordon <16200219+pierlon@users.noreply.github.com>

johnwatkins0 committed:
Rename MonitorCssTransientCachingTest

johnwatkins0 committed:
Merge branch 'feature/1756-validation-cron' of https://github.com/ampproject/amp-wp into feature/1756-validation-cron

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

pierlon reviewed (approved):
Well done :100:.

westonruter commented on includes/cli/class-amp-cli-validation-command.php:
```diff
@@ -248,9 +240,11 @@ private function get_validation_url_provider() {
 		$this->scannable_url_provider = new ScannableURLProvider(
-			$limit_type_validate_count,
-			$include_conditionals,
-			$force_crawl_urls
+			new URLScanningContext(
```
This can be addressed later, but shouldn't both `ScannableURLProvider` and `URLScanningContext` be instantiated via the injector?

johnwatkins0 replied:
Yes, I have refrained so far from moving this class into the DI architecture, but it could be done easily.

westonruter commented on src/AmpWpPlugin.php:
```diff
@@ -9,8 +9,11 @@
 
 use AmpProject\AmpWP\Admin;
 use AmpProject\AmpWP\BackgroundTask;
+use AmpProject\AmpWP\Infrastructure\Injector;
```
This `use` isn't being used.

```suggestion
```

westonruter commented on src/BackgroundTask/MonitorCssTransientCaching.php:
```diff
@@ -92,23 +92,25 @@ protected function get_event_name() {
-		if ( null === $date ) {
-			$date = new DateTimeImmutable();
-		}
+		/* @var DateTimeInterface */
+		$date = isset( $args[0] ) ? $args[0] : new DateTimeImmutable();
```
```suggestion
		$date = isset( $args[0] ) && $args[0] instanceof DateTimeInterface ? $args[0] : new DateTimeImmutable();
```

westonruter commented on src/BackgroundTask/MonitorCssTransientCaching.php:
```diff
@@ -92,23 +92,25 @@ protected function get_event_name() {
-		if ( null === $transient_count ) {
-			$transient_count = $this->query_css_transient_count();
-		}
+		/** @var int */
+		$transient_count = isset( $args[1] ) ? $args[1] : $this->query_css_transient_count();
```
```suggestion
		$transient_count = isset( $args[1] ) ? (int) $args[1] : $this->query_css_transient_count();
```

westonruter commented on src/BackgroundTask/RecurringBackgroundTask.php:
```diff
@@ -0,0 +1,55 @@
+	 * @param mixed[] ...$args Arguments passed to the function from the action hook.
+	 */
+	public function schedule_event( ...$args ) {
+		if ( ! current_user_can( 'manage_options' ) ) {
+			return;
+		}
```
Why the `manage_options` check? Is this to prevent scheduling for frontend requests. Yeah, we've done that before.

westonruter replied:
Perhaps it would be safer to gate this behind just a user being logged-in:

```suggestion
		if ( ! is_user_logged_in() ) {
			return;
		}
```

johnwatkins0 replied:
Yes, this works.

westonruter commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -0,0 +1,142 @@
+			return false;
+		}
+
+		if ( ! current_user_can( 'manage_options' ) ) {
+			return false;
+		}
+
```
Won't the `manage_options` check mean that non-admins who may have DevTools enabled won't get this scheduled action? In any case, is not this redundant with the `is_user_enabled()` check immediately below?

```suggestion
```

westonruter commented on src/Validation/SavePostValidationEvent.php:
```diff
@@ -0,0 +1,142 @@
+			return false;
+		}
+
+		// Validation is performed on post save if user has dev tools on.
+		if ( ! $this->dev_tools_user_access->is_user_enabled( wp_get_current_user() ) ) {
```
Is this logic not inverted? If `is_user_enabled()` returns false, then DevTools is disabled, and then that means we should go ahead with scheduling the event.

```suggestion
		// Validation is performed on post save if user has dev tools on.
		if ( $this->dev_tools_user_access->is_user_enabled( wp_get_current_user() ) ) {
```

So, if the user has dev tools turned on, we should short-circuit because validation will be performed synchronously (at the moment).

johnwatkins0 replied:
Yes, you're right.

westonruter commented on src/Validation/ScannableURLProvider.php:
```diff
@@ -87,8 +61,9 @@ public function get_urls( $offset = 0 ) {
 		);
 		$public_post_types      = get_post_types( [ 'public' => true ] );
 
+		$limit_per_type = $this->context->get_limit_per_type();
```
```suggestion
		$limit_per_type         = $this->context->get_limit_per_type();

```

westonruter commented on src/Validation/URLScanningContext.php:
```diff
@@ -0,0 +1,106 @@
+/**
+ * URLScanningContext class.
+ *
+ * @since 2.1
```
```suggestion
 * @since 2.1
 *
 * @internal
```

westonruter commented on src/Validation/URLScanningContext.php:
```diff
@@ -0,0 +1,106 @@
+	/**
+	 * Class constructor.
+	 *
+	 * @param integer $limit_per_type       The maximum number of URLs to validate for each type.
+	 * @param array   $include_conditionals An allowlist of conditionals to use for validation.
+	 * @param boolean $include_unsupported  Whether to include URLs that don't support AMP.
```
```suggestion
	 * @param int   $limit_per_type       The maximum number of URLs to validate for each type.
	 * @param array $include_conditionals An allowlist of conditionals to use for validation.
	 * @param bool  $include_unsupported  Whether to include URLs that don't support AMP.
```

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,119 @@
+		 * @since 2.1
+		 * @param int The number of seconds. Default 1.
+		 */
+		return absint( apply_filters( 'amp_url_validation_sleep_time', self::DEFAULT_SLEEP_TIME ) );
```
```suggestion
		return max( (int) apply_filters( 'amp_url_validation_sleep_time', self::DEFAULT_SLEEP_TIME ), 0 );
```

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,119 @@
+		 * Filters the length of time to sleep between validating URLs.
+		 *
+		 * @since 2.1
+		 * @param int The number of seconds. Default 1.
```
```suggestion
		 * @param int The number of seconds. Default 1. Setting to 0 or a negative numbers disables all throttling.
```

johnwatkins0 committed:
Remove unneeded use statement

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Improve how potentially empty variables are set

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Schedule event if user is logged in

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Remove unneeded manage_options check

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Fix reversed logic of dev_tools_user_access check

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Remove blank line between assignments

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Mark class as internal

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Fix type comments

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Make amp_url_validation_sleep_time have a minimum of 0

Co-authored-by: Weston Ruter <westonruter@google.com>

johnwatkins0 committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

johnwatkins0 committed:
Fix failing PHPUnit

johnwatkins0 committed:
Reorder patches in composer.json

johnwatkins0 committed:
Regenerate composer.lock

johnwatkins0 committed:
Reorder composer.json fields

johnwatkins0 committed:
Update param type for phpstan

johnwatkins0 committed:
Use PHPStan ignore for 'cannot cast array to int'

westonruter committed:
Merge remote-tracking branch 'origin/develop' into feature/1756-validation-cron

* origin/develop:
  Run composer update
  Ensure alphabetical order for patches in composer.json
  Reorder composer.json fields
  Remove obsolete amp_is_canonical() examples
  Add more markdown formatting and restore linebreak
  Wrap HTML tag in backticks
  Add variable to docs for `amp_dev_mode_enabled` filter
  Fix parsing of docblocks that contain code snippets

westonruter committed:
Update service name for SavePostValidationEvent and add to PHPSTORM_META

westonruter commented on src/BackgroundTask/RecurringBackgroundTask.php:
```diff
@@ -0,0 +1,55 @@
+		}
+
+		$event_name = $this->get_event_name();
+		$timestamp  = wp_next_scheduled( $event_name );
```
I found that a `amp_validate_urls` event was getting registered with every request, and the reason is that `wp_next_scheduled( $event_name )` was never returning a timestamp. This is due to the inclusion of `$args` in 1419ef4e3ccb4654d6e6f70e8ddfa19c8e8f68da which also needed to be passed to `wp_next_scheduled()`:

```suggestion
		$timestamp  = wp_next_scheduled( $event_name, $args );
```

westonruter committed:
Add missing $args to wp_next_scheduled() call after 1419ef4

westonruter reviewed (changes requested):
@johnwatkins0 I tried turning off DevTools to see the `SingleScheduledBackgroundTask` in action. However, I found that `SavePostValidationEvent::should_schedule_event()` would never get called. The reason appears to be the logic in `is_needed`:

```php
	public static function is_needed() {
		return is_admin() || wp_doing_cron();
	}
```

When making an edit to a post in the block editor, neither `is_admin()` nor `wp_doing_cron()` are true for the REST API request. This means the `save_post` action is never added.

Should the `CronBasedBackgroundTask` abstract class not be made `Conditional`?

johnwatkins0 committed:
Make CronBasedBackgroundTask non-Conditional

johnwatkins0 committed:
Merge branch 'feature/1756-validation-cron' of https://github.com/ampproject/amp-wp into feature/1756-validation-cron

johnwatkins0 committed:
Remove obsolete test assertion

johnwatkins0 commented:
> Should the `CronBasedBackgroundTask` abstract class not be made `Conditional`?

That's the simple solution. Updated in a53afa5

westonruter committed:
Reduce URLValidationCron frequency to daily

westonruter commented on src/Validation/URLValidationCron.php:
```diff
@@ -0,0 +1,119 @@
+		$this->url_validation_provider = $url_validation_provider;
+	}
+
+	/**
+	 * Callback for the cron action.
+	 *
+	 * @param mixed[] ...$args Unused callback arguments.
+	 */
+	public function process( ...$args ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		$urls       = $this->scannable_url_provider->get_urls();
+		$sleep_time = $this->get_sleep_time();
+
+		foreach ( $urls as $url ) {
+			$this->url_validation_provider->get_url_validation( $url['url'], $url['type'], true );
+			if ( $sleep_time ) {
+				sleep( $sleep_time );
+			}
+		}
+	}
```
We can address this in a subsequent PR, but I have an idea for how this could be modified to not make use of `sleep()`. Instead of there being one single event that scans all of the URLs, we can instead register a separate event for each for each URL, and offset the `time()` for each event by 10 minute intervals (for example). In this way only one URL would be validated during any given cron request, and there would be no need for `sleep()`.

westonruter replied:
See #5750.

westonruter committed:
Fix ability for BackgroundTaskDeactivator to unschedule hooks

westonruter commented:
I was testing `BackgroundTaskDeactivator` and I found it not to be working. Given scheduled AMP hooks as follows:

```
$ wp cron event list | grep amp
| amp_validated_url_stylesheet_gc    | 2020-12-30 05:15:19 | 59 minutes 25 seconds | 1 hour        |
| amp_monitor_css_transient_caching  | 2020-12-31 04:15:19 | 23 hours 59 minutes   | 1 day         |
| amp_validate_urls                  | 2020-12-31 04:15:19 | 23 hours 59 minutes   | 1 day         |
```

After deactivating, they remained. I found there were a few problems:

1. `BackgroundTaskDeactivator` was not in the `SERVICES` array so it was not getting `register`'ed.
2. The service was `Conditional` based on `is_admin() || wp_doing_cron()` which meant it was not run when deactivating with WP-CLI.
3. With the passing of `$args` to scheduled events in 1419ef4e3ccb4654d6e6f70e8ddfa19c8e8f68da, this broke the use of `wp_clear_scheduled_hook()` which requires passing the same `$args`. In reality, the correct function here to use was `wp_unschedule_hook()` which unregisters all hooks regardless of the args.

I've addressed these points in f42473590b29fbf5b7434c440b8ddddc5b5a60be. Now when deactivating the plugin, no AMP hooks remain:

```
$ wp cron event list | grep amp
| amp_monitor_css_transient_caching  | 2020-12-30 04:22:13 | now                   | 1 day         |
| amp_validated_url_stylesheet_gc    | 2020-12-30 04:22:13 | now                   | 1 hour        |
| amp_validate_urls                  | 2020-12-30 04:22:13 | now                   | 1 day         |

$ wp plugin deactivate amp
Plugin 'amp' deactivated.
Success: Deactivated 1 of 1 plugins.

$ wp cron event list | grep amp

```

westonruter committed:
Make RecurringBackgroundTask::schedule_event() final

westonruter committed:
Guard against empty post ID causing global post to be returned by get_post()

westonruter committed:
Mark URLValidationProvider and ScannableURLProvider as internal

westonruter committed:
Ensure amp_url_validation_limit_per_type filter returns int

westonruter committed:
Simplify logic to ensure -1 <= limit_per_type <= PHP_INT_MAX

westonruter reviewed (approved):
Thanks for the mammoth effort here. This is going to go a long way to help site owners monitor their sites for AMP validation issues.
